src/persist/src/file.rs

```diff
@@ -0,0 +1,197 @@
+            bytes.read_exact(&mut buf)?;
+            let val_len = u64::from_le_bytes(buf);
+            let mut val = vec![0u8; val_len.try_into().unwrap()];
+            bytes.read_exact(&mut val[..])?;
+            let val = String::from_utf8(val).expect("failed to pass val");
+
+            // Decode time
+            bytes.read_exact(&mut buf)?;
+            let ts = u64::from_le_bytes(buf);
+
+            // Decode diff
+            bytes.read_exact(&mut buf)?;
+            let diff = i64::from_le_bytes(buf);
+
+            dataz.push(((key, val), ts, diff as isize));
+        }
+        Ok(MemSnapshot { dataz })
+    }
+
+    fn allow_compaction(&mut self, _ts: u64) -> Result<(), Error> {
+        // No-op for now.
+        Ok(())
+    }
+}
+
+/// Implementation of [Persister]
+pub struct FilePersister {
+    base_path: PathBuf,
+    registered: HashSet<Id>,
+    dataz: HashMap<Id, FileStream>,
+}
+
+impl FilePersister {
+    /// Constructs a new, empty SimplePersister.
+    pub fn new<P: AsRef<Path>>(base_path: P) -> Self {
+        std::fs::create_dir_all(&base_path).expect("failed to create FilePersister base path");
+        FilePersister {
+            base_path: base_path.as_ref().to_path_buf(),
+            registered: HashSet::new(),
+            dataz: HashMap::new(),
+        }
+    }
+}
+
+impl Persister for FilePersister {
+    type Write = FileStream;
+    type Meta = FileStream;
+
+    fn create_or_load(&mut self, id: Id) -> Result<Token<Self::Write, Self::Meta>, Error> {
+        if self.registered.contains(&id) {
+            return Err(format!("internal error: {:?} already registered", id).into());
+        }
+
+        self.registered.insert(id);
+        let path = self.base_path.join(format!("file-persister-{}", id.0));
+        let p = self
+            .dataz
+            .entry(id)
+            .or_insert_with(|| FileStream::new(path));
+        let t = Token {
+            write: p.clone(),
+            meta: p.clone(),
+        };
+        Ok(t)
+    }
+
+    // Note that we don't remove from registered here and maybe we ought to?
+    fn destroy(&mut self, id: Id) -> Result<(), Error> {
+        if self.dataz.remove(&id).is_none() {
+            return Err(format!("internal error: {:?} not registered", id).into());
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_file_stream() {
+        let data = vec![
+            (("key1".to_string(), "val1".to_string()), 1, 1),
+            (("key2".to_string(), "val2".to_string()), 1, 1),
+        ];
+
+        let mut file_stream = FileStream::new("test".to_string());
+        file_stream.write_sync(&data).expect("fixme");
+        let snapshot = file_stream.snapshot().expect("fixme");
+        let input_snapshot = MemSnapshot { dataz: data };
+        assert_eq!(snapshot, input_snapshot);
+
+        // Cleanup test state
+        std::fs::remove_file("test").expect("failed to remove test state");
```

Comment:
make this some path in the temp dir (`std::env::temp_dir`) so even if this cleanup doesn't run (failed tests just panic so it often wont) the file isn't orphaned. preferably something like `[temp_dir]`/`[nanosecond "now" timestamp] so multiple runs of the test don't conflict. go had something that automated all this for you, dunno if rust has something similar

Reply:
done -- used temp_dir and removed the explicit cleanup. I still think its maybe a little bit useful to be able to introspect the file on test failure, but its easy enough to comment out the temp directory bit if we need to look at the file